spotify_api.rb

```diff
@@ -0,0 +1,112 @@
+require 'httparty'
+require 'pry'
+require 'json'
+require 'base64'
+require './spotify_api_token'
+
+Spotify_api = "https://api.spotify.com"
+
+class SpotifyApiRequest
+
+  attr_reader :song, :raw_data, :token
+
+  def initialize authorization:, song:, test_file: nil
+    @song = song
+    @authorization = authorization
+    @raw_data = []
+    @test_file = test_file
+  end
+
+  def get_song_query
+    st_encoded = URI.encode @song
+    HTTParty.get(
+      Spotify_api + "/v1/search?q=#{st_encoded}&type=#{@type}",
+      headers: { "Accept" => "application/json", "Authorization" => @authorization }
+    )
+  end
+
+  def export_playlist playlist
+    playlist.spotify_id = create_playlist
+    playlist.save!
+    export_songs_to_playlist playlist
+  end
+
+  def parse!
+    if @song
+         @type = "track"
+
+      # branch to @type assignment for more user-friendly, and optimal search ?
+
+      #elsif album: && artist:
+        #@type =
+    end
+
+    if @test_file
+      @raw_data = JSON.parse(File.read @test_file)
+    else
+      @raw_data = get_song_query
+    end
+  end
+
+
+  def get_songs
+    return [] if raw_data == []
+
+    raw_songs = []
+    @raw_data["tracks"]["items"].each do |item|
+      raw_songs.push(item)
+    end
+
+    each_song_array = []
+    raw_songs.each do |song|
+      song_hash = {}
+      song.keys.each do |key|
+        case key
+        when "album"
+          song_hash[key] = song[key]["name"]
+        when "artists"
+          song_hash["artist"] = song[key].first["name"]
+          song["artists"].map { |h| h["name"] }.join(", ")
+        when "id"
+          song_hash["id"] = song[key]
+        when "name"
+          song_hash["title"] = song[key]
+        end
+      end
+      each_song_array.push(song_hash)
+    end
+    each_song_array
+  end
+
+  private # --- Everything from here down is only callable from this object ----
+
+  def create_playlist
+    r = HTTParty.post(
+      Spotify_api + "/v1/users/ferretpenguin/playlists",
+      headers: { "Accept" => "application/json", "Authorization" => @authorization},
```

Comment:
There's a fair amount of duplication in all the various endpoint calls. HTTParty does have provisions for setting a `base_uri` and default `headers` that would be worth looking in to here (as much so readers can focus on the specifics of each endpoint when reading them)